platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r89/ToolingApiPolymorphismCrossVersionTest.groovy

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.tooling.r89
+
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.integtests.tooling.fixture.CompositeModel
+import org.gradle.integtests.tooling.fixture.DeepChildModel
+import org.gradle.integtests.tooling.fixture.ShallowChildModel
+import org.gradle.integtests.tooling.fixture.SideModel
+import org.gradle.integtests.tooling.fixture.ToolingApiSpecification
+import org.gradle.integtests.tooling.fixture.ToolingApiVersion
+import org.gradle.integtests.tooling.fixture.BaseModel
+import org.gradle.tooling.provider.model.ToolingModelBuilder
+import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry
+
+import javax.inject.Inject
+
+@ToolingApiVersion(">=8.9")
+class ToolingApiPolymorphismCrossVersionTest extends ToolingApiSpecification {
+
+    def "setup"() {
+        file("plugins/build.gradle") << """
+            plugins {
+                id("groovy-gradle-plugin")
+            }
+            gradlePlugin {
+                plugins {
+                    test {
+                        id = "my.plugin"
+                        implementationClass = "my.MyPlugin"
+                    }
+                }
+            }
+        """
+
+        file("plugins/src/main/groovy/my/MyModel.groovy") << """
+            package org.gradle.integtests.tooling.fixture
+
+            interface BaseModel {
+            }
+
+            interface ShallowChildModel extends BaseModel {
+                String getShallowMessage()
+            }
+
+            interface DeepChildModel extends BaseModel {
+                String getDeepMessage()
+            }
+
+            interface SideModel extends BaseModel {
+            }
+
+            abstract class AbstractModel implements DeepChildModel {
+            }
+
+            class DefaultModel extends AbstractModel implements ShallowChildModel, SideModel, java.io.Serializable {
+                private final String message
+                DefaultModel(String message) { this.message = message }
+                String getShallowMessage() { "shallow " + message }
+                String getDeepMessage() { "deep " + message }
+                String toString() { message }
```

Comment:
Maybe add `@Override` annotations here for clarity.

Reply:
Done.